src/Snail/Lexer.hs

```diff
@@ -1,50 +1,99 @@
-  skipBlockComment,
-  skipLineComment,
-  spaces,
-  symbol,
+  -- * Character lists
+  initialCharacter,
+  specialInitialCharacter,
+  digitCharacter,
+  specialSubsequentCharacter,
+
+  -- * The only parser you'll ever need
+  SExpression (..),
+  sExpression,
 ) where
 
+import Data.Functor (($>))
 import Data.Text (Text)
+import Data.Text qualified as Text
 import Data.Void
-import Text.Megaparsec
+import Text.Megaparsec hiding (Tokens (..))
 import Text.Megaparsec.Char
 import Text.Megaparsec.Char.Lexer qualified as L
 
 -- | TODO: 'Void' is the error type but we should use an explicit error type
 type Parser = Parsec Void Text
 
--- | Megaparsec's 'skipLineComment' takes a prefix and skips lines that begin with that prefix
+{- | Megaparsec's 'skipLineComment' takes a prefix and skips lines that begin
+ with that prefix
+-}
 skipLineComment :: Parser ()
 skipLineComment = L.skipLineComment ";"
 
--- | Megaparsec's 'skipBlockComment' takes prefix and suffix and skips anything in between
+{- | Megaparsec's 'skipBlockComment' takes prefix and suffix and skips anything
+ in between
+-}
 skipBlockComment :: Parser ()
 skipBlockComment = L.skipBlockCommentNested "#|" "|#"
 
--- | Generate a parser for whitespace in a language with 'skipLineComment' and 'skipBlockComment'
+{- | Generate a parser for whitespace in a language with 'skipLineComment' and
+ 'skipBlockComment'
+-}
 spaces :: Parser ()
 spaces = L.space space1 skipLineComment skipBlockComment
 
--- | A 'lexeme' is ...
-lexeme :: Parser a -> Parser a
-lexeme = L.lexeme spaces
-
 -- | Parse a 'Text' verbatim
 symbol :: Text -> Parser Text
 symbol = L.symbol spaces
 
--- | Parse an 'Integer' as a lexeme
-integer :: Parser Integer
-integer = lexeme L.decimal
-
--- | Parse a signed 'Integer'
-signedInteger :: Parser Integer
-signedInteger = L.signed spaces integer
-
 -- | Parse an S-Expression
 parens :: Parser a -> Parser a
 parens = between (symbol "(") (symbol ")")
+
+-- | A 'lexeme' is ...
+lexeme :: Parser a -> Parser a
+lexeme = L.lexeme spaces
+
+-- | ...
+initialCharacter :: String
+initialCharacter = ['a' .. 'z'] <> ['A' .. 'Z']
+
+-- | ...
+specialInitialCharacter :: String
+specialInitialCharacter = "!$%&*/:<=>?^_~"
+
+-- | ...
+digitCharacter :: String
+digitCharacter = ['0' .. '9']
+
+-- | ...
+specialSubsequentCharacter :: String
+specialSubsequentCharacter = "+-.@"
+
+-- | The list of valid token characters, note that we allow invalid tokens at this point
+validCharacter :: Parser Char
+validCharacter =
+  oneOf
+    ( initialCharacter
+        <> specialInitialCharacter
+        <> digitCharacter
+        <> specialSubsequentCharacter
+    )
+
+-- | ...
+data SExpression
+  = Token (SourcePos, Text)
+  | SExpression [SExpression]
+  deriving (Eq, Show)
+
+-- | ...
+term :: Parser SExpression
+term = text <|> sExpression
+
+-- | ...
+text :: Parser SExpression
+text = do
+  token <- some validCharacter
+  sourcePos <- getSourcePos
+  pure $ Token (sourcePos, Text.pack token)
```

Comment:
TODO: subtract the length of the token from the `sourcePos`

Reply:
couldn't you just get the source position before you get the token? or is this to handle unicode length weirdness

Reply:
There is a comment here https://hackage.haskell.org/package/megaparsec-9.2.0/docs/Text-Megaparsec.html#v:getSourcePos that suggests this is very expensive. So, I want to avoid calling it if `some validCharacter` fails.

Reply:
That being said, I think it isn't worth doing because implementing the subtraction adds unnecessary read overhead.